src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,282 @@
+            String sessionId) throws WebAuthnAuthenticationException {
+        try {
+            AssertionRequest assertionRequest = activeAuthentications.get(sessionId);
+            AssertionResult result = rp.finishAssertion(FinishAssertionOptions.builder().request(assertionRequest)
+                    // The PublicKeyCredentialRequestOptions from startAssertion above
+                    .response(pkc).build());
+            if (result.isSuccess() && result.isSignatureCounterValid()) {
+                final WebAuthnUserAccount account = webAuthnUserAccountRepository
+                        .findByUserHandle(result.getUserHandle().getBase64());
+                List<WebAuthnCredential> credentials = webAuthnCredentialsRepository
+                        .findByParentAccountId(account.getId());
+                Optional<WebAuthnCredential> toUpdate = Optional.empty();
+                ByteArray resultCredentialId = result.getCredentialId();
+                for (WebAuthnCredential c : credentials) {
+                    ByteArray cCredentialId = ByteArray.fromBase64(c.getCredentialId());
+                    if (cCredentialId.equals(resultCredentialId)) {
+                        toUpdate = Optional.of(c);
+                    }
+                }
+                if (toUpdate.isEmpty()) {
+                    throw new WebAuthnAuthenticationException(account.getSubject(),
+                            "Could not find the requested credential in the account");
+                }
+                WebAuthnCredential credential = toUpdate.get();
+                credentials.remove(credential);
+                credential.setSignatureCount(result.getSignatureCount());
+                credential.setLastUsedOn(new Date());
+                webAuthnCredentialsRepository.save(credential);
+                final String username = account.getUsername();
+                if (!StringUtils.hasText(username)) {
+                    throw new WebAuthnAuthenticationException(account.getSubject(),
+                            "Could not find the requested credential in the account");
+                }
+                return username;
+            } else {
+                return null;
+            }
+        } catch (Exception e) {
+            throw new WebAuthnAuthenticationException("_",
+                    "Login failed");
+        }
+    }
+
+    UserIdentity getUserIdentityOrGenerate(String username, String realm, String displayName,
+            Subject subjectOrNull) {
+        final String userDisplayName = displayName != null ? displayName : "";
+        Optional<UserIdentity> option = getUserIdentity(username, userDisplayName);
+        if (option.isPresent()) {
+            return option.get();
+        } else {
+            byte[] userHandle = new byte[64];
+            SecureRandom random = new SecureRandom();
+            random.nextBytes(userHandle);
```

Comment:
user handle generation deserves a custom private method `generateUserHandle`

Reply:
use an existing key generator for this instead of manually fetching random bytes.
For example put as class prop
```
     private final StringKeyGenerator keyGenerator = new Base64StringKeyGenerator(Base64.getUrlEncoder());
```
and then call
```
     String userHandle = keyGenerator.generateKey();
```

look here 
https://github.com/scc-digitalhub/AAC/blob/7e137c8ca2a6826168d02efa95c32b298dce6688/src/main/java/it/smartcommunitylab/aac/spid/auth/SpidWebSsoAuthenticationRequestFilter.java#L196